Implement scenes tiles in TileMaps

groud opened the pull request:
![Peek 18-05-2021 13-28](https://user-images.githubusercontent.com/6093119/118672816-d1872700-b7f8-11eb-918d-4c32934689e5.gif)

Implements a TileSetScenesCollectionSource for TileSets. Each tile in this TileSet source has a scene that will be instantiated as an internal child of the TileMap node.

I also implemented a "display placeholder" property that allow displaying a placeholder at the position of the tile in case a scene is not visible.

Also, I moved `TileSetAtlasSource::INVALID_*` to `TileSetSource::INVALID_*`

groud committed:
Implement scenes tiles in TileMaps

Calinou commented on scene/resources/tile_set.cpp:
```diff
@@ -2034,6 +2035,203 @@ void TileSetAtlasSource::_compute_next_alternative_id(const Vector2i p_atlas_coo
+	int new_scene_id = p_id_override >= 0 ? p_id_override : next_scene_id;
+
+	scenes[new_scene_id] = SceneData();
+	;
```
There is a lone semicolon:

```suggestion
```

Calinou commented on doc/classes/TileSetScenesCollectionSource.xml:
```diff
@@ -0,0 +1,155 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="TileSetScenesCollectionSource" inherits="TileSetSource" version="4.0">
+	<brief_description>
```
Remember to add documentation for this class :slightly_smiling_face:

groud replied:
Yeah, I'll do a documentation pass once I'm sure the API is in its final form. For now I only synchronize the docs.

Calinou commented on editor/plugins/tiles/tile_map_editor.cpp:
```diff
@@ -1746,6 +1890,28 @@ TileMapEditorTilesPlugin::TileMapEditorTilesPlugin() {
+	scene_tiles_list->set_select_mode(ItemList::SELECT_MULTI);
+	scene_tiles_list->connect("multi_selected", callable_mp(this, &TileMapEditorTilesPlugin::_scenes_list_multi_selected));
+	scene_tiles_list->connect("nothing_selected", callable_mp(this, &TileMapEditorTilesPlugin::_scenes_list_nothing_selected));
+	/*scene_tiles_list->connect("item_activated", callable_mp(this, &TileSetScenesCollectionSourceEditor::_scenes_list_item_activated));*/
```
Remove commented out code (unless you meant to uncomment it):

```suggestion
```

Calinou commented on editor/plugins/tiles/tile_set_scenes_collection_source_editor.cpp:
```diff
@@ -0,0 +1,514 @@
+	if (tile_set_scenes_collection_source) {
+		tile_set_scenes_collection_source->disconnect("changed", callable_mp(this, &TileSetScenesCollectionSourceEditor::_tile_set_scenes_collection_source_changed));
+	}
+
+	// Clear the selection.
+	// selection.clear();
```
Remove commented out code unless you intend to uncomment it:

```suggestion
```

Calinou commented on scene/resources/tile_set.h:
```diff
@@ -636,6 +682,18 @@ class TileSetAtlasPluginNavigation : public TileSetPlugin {
+
+public:
+	// Tilemap updates
+	//virtual void tilemap_notification(TileMap *p_tile_map, int p_what) override;
```
Commented out code:

```suggestion
```

Calinou commented on scene/resources/tile_set.h:
```diff
@@ -636,6 +682,18 @@ class TileSetAtlasPluginNavigation : public TileSetPlugin {
+	// Tilemap updates
+	//virtual void tilemap_notification(TileMap *p_tile_map, int p_what) override;
+	virtual void update_dirty_quadrants(TileMap *p_tile_map, SelfList<TileMapQuadrant>::List &r_dirty_quadrant_list) override;
+	//virtual void create_quadrant(TileMap *p_tile_map, TileMapQuadrant *p_quadrant) override;
```
Commented out code:

```suggestion
```

KoBeWi commented:
- Not sure how you should prevent that, but it's possible to draw the scene that you are currently editing, resulting in cyclic dependency. Although it just won't save, so maybe it's handled correctly.
- I managed to make the editor crash once, not sure how:
```
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] Ref<SceneState>::operator-> (C:\godot_source\core\object\reference.h:109)
[1] PackedScene::instance (C:\godot_source\scene\resources\packed_scene.cpp:1626)
[2] TileSetPluginScenesCollections::update_dirty_quadrants (C:\godot_source\scene\resources\tile_set.cpp:4484)
[3] TileMap::update_dirty_quadrants (C:\godot_source\scene\2d\tile_map.cpp:373)
[4] call_with_variant_args_helper<TileMap> (C:\godot_source\core\variant\binder_common.h:214)
[5] call_with_variant_args_dv<TileMap> (C:\godot_source\core\variant\binder_common.h:357)
[6] MethodBindT<TileMap>::call (C:\godot_source\core\object\method_bind.h:285)
[7] Object::call (C:\godot_source\core\object\object.cpp:784)
[8] Callable::call (C:\godot_source\core\variant\callable.cpp:53)
[9] MessageQueue::_call_function (C:\godot_source\core\object\message_queue.cpp:259)
[10] MessageQueue::flush (C:\godot_source\core\object\message_queue.cpp:306)
[11] SceneTree::physics_process (C:\godot_source\scene\main\scene_tree.cpp:415)
[12] Main::iteration (C:\godot_source\main\main.cpp:2479)
[13] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:622)
[14] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:163)
[15] _main (C:\godot_source\platform\windows\godot_windows.cpp:185)
[16] main (C:\godot_source\platform\windows\godot_windows.cpp:199)
[17] __scrt_common_main_seh (D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[18] BaseThreadInitThunk
-- END OF BACKTRACE --
```
- Scene tiles are always centered. They either should have configurable offset or at least use the position of instance's root node.
- I get this bug for some reason:
![le Bug](https://user-images.githubusercontent.com/2223172/118802323-47939880-b8a2-11eb-8ba4-f9d6c1cf50c4.gif)

groud commented:
> Not sure how you should prevent that, but it's possible to draw the scene that you are currently editing, resulting in cyclic dependency. Although it just won't save, so maybe it's handled correctly.

I don't think I can do anything about it. TileSet is a resource, so it can't be strictly associated to the scene it is in. So well... as soon as it does not crashes I believe this is fine, as the engine does not checks for cyclical dependencies in general.

> I managed to make the editor crash once, not sure how:

It happens when trying to paint a scene with an invalid scene. It should be fixed now.

> Scene tiles are always centered. They either should have configurable offset or at least use the position of instance's root node.

Fixed.

> I get this bug for some reason:

Fixed too, I screwed things up when rebasing... ^^

groud commented:
I have no clue why the Android build fails and how to solve it. That's annoying... :/

KoBeWi commented:
Seems like it's possible to draw 2 scenes at one when you click between tiles:
![Bug2](https://user-images.githubusercontent.com/2223172/118880390-1773e600-b8f2-11eb-9031-f05cb11a6818.gif)
But only horizontally.

groud commented:
> Seems like it's possible to draw 2 scenes at one when you click between tiles:

Oh oh, that was a tricky one. My _draw_line function took one Vector2i and one Vector2 as arguments, while they should have been both Vector2. This caused very small differences between the two computed ends of a line, even if, in that case, the end and the beginning of the line should were supposed to have the same value.

Anyway, that's fixed now, thanks!

akien-mga reviewed (approved):
I didn't test, but since @KoBeWi did and code seems fine, should be good to go.
